File: src/main/java/pt/joaomneto/titancompanion/adventurecreation/impl/MRAdventureCreation.java
package pt.joaomneto.titancompanion.adventurecreation.impl;

import android.view.View;

import java.io.BufferedWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import pt.joaomneto.titancompanion.R;
import pt.joaomneto.titancompanion.adventure.Adventure.AdventureFragmentRunner;
import pt.joaomneto.titancompanion.adventure.impl.fragments.mr.MRSkill;
import pt.joaomneto.titancompanion.adventure.impl.util.Spell;
import pt.joaomneto.titancompanion.adventurecreation.impl.fragments.mr.MRAdventureCreationSkillsFragment;

public class MRAdventureCreation extends TWOFMAdventureCreation {

    private final static int FRAGMENT_MR_SPELLS = 1;
    private final static int FRAGMENT_MR_POTION = 2;

    List<Spell> chosenSkills = new ArrayList<>();

    public MRAdventureCreation() {
        super();
        Companion.getFragmentConfiguration().clear();
        Companion.getFragmentConfiguration().put(0, new AdventureFragmentRunner(
                R.string.title_adventure_creation_vitalstats,
                "pt.joaomneto.titancompanion.adventurecreation.impl.fragments.VitalStatisticsFragment"));
        Companion.getFragmentConfiguration().put(FRAGMENT_MR_SPELLS, new AdventureFragmentRunner(
                R.string.specialSkills,
                "pt.joaomneto.titancompanion.adventurecreation.impl.fragments.mr.MRAdventureCreationSkillsFragment"));
        Companion.getFragmentConfiguration().put(FRAGMENT_MR_POTION, new AdventureFragmentRunner(R.string.title_adventure_creation_potions,
                "pt.joaomneto.titancompanion.adventurecreation.impl.fragments.PotionsFragment"));

    }

    @Override
    protected void rollGamebookSpecificStats(View view) {

    }

    @Override
    protected void storeAdventureSpecificValuesInFile(BufferedWriter bw)
            throws IOException {

        String chosenSkillsS = "";

        if (!chosenSkills.isEmpty()) {
            for (Spell spell : getSkills()) {
                chosenSkillsS += spell + "#";
            }
            chosenSkillsS = chosenSkillsS.substring(0, chosenSkillsS.length() - 1);
        }

        bw.write("skills=" + chosenSkillsS + "\n");
        bw.write("provisionsValue=4\n");
        bw.write("provisions=10\n");
        bw.write("gold=0\n");
    }

    @Override
    public String validateCreationSpecificParameters() {
        StringBuilder sb = new StringBuilder();
        boolean error = false;
        if (this.chosenSkills == null || this.chosenSkills.isEmpty()) {
            sb.append(getString(R.string.chosenSkills));
        }

        return sb.toString();
    }

    private MRAdventureCreationSkillsFragment getMRSkillFragment() {
        MRAdventureCreationSkillsFragment mrSkillFragment = (MRAdventureCreationSkillsFragment) Companion.getFragments().get(FRAGMENT_MR_SPELLS);
        return mrSkillFragment;
    }


    public synchronized int getSpellValue() {
        return 3;
    }

    public void addSkill(MRSkill skill) {
        if (!chosenSkills.contains(skill)) {
            chosenSkills.add(skill);
        }
    }

    public void removeSkill(int skill) {
        chosenSkills.remove(skill);
    }

    public List<Spell> getSkills() {
        return chosenSkills;
    }
}

File: src/main/java/pt/joaomneto/titancompanion/adventurecreation/impl/MRAdventureCreation.kt
//package pt.joaomneto.titancompanion.adventurecreation.impl
//
//import pt.joaomneto.titancompanion.R
//import pt.joaomneto.titancompanion.adventure.impl.fragments.mr.MRSkill
//import pt.joaomneto.titancompanion.adventure.impl.util.Spell
//import pt.joaomneto.titancompanion.adventurecreation.impl.fragments.PotionsFragment
//import pt.joaomneto.titancompanion.adventurecreation.impl.fragments.VitalStatisticsFragment
//import pt.joaomneto.titancompanion.adventurecreation.impl.fragments.mr.MRAdventureCreationSkillsFragment
//import pt.joaomneto.titancompanion.util.AdventureFragmentRunner
//import java.io.BufferedWriter
//import java.io.IOException
//import java.util.ArrayList
//
//class MRAdventureCreation : TWOFMAdventureCreation(
//    arrayOf(
//        AdventureFragmentRunner(
//            R.string.title_adventure_creation_vitalstats,
//            VitalStatisticsFragment::class
//        ),
//        AdventureFragmentRunner(
//            R.string.title_adventure_creation_potions,
//            PotionsFragment::class
//        ),
//        AdventureFragmentRunner(
//            R.string.specialSkills,
//            MRAdventureCreationSkillsFragment::class
//        )
//    )
//) {
//
//    val skills: MutableList<Spell> = ArrayList()
//
//    val spellValue: Int = 3
//
//    @Throws(IOException::class)
//    override fun storeAdventureSpecificValuesInFile(bw: BufferedWriter) {
//
//        var chosenSkillsS = ""
//
//        if (!skills.isEmpty()) {
//            for (spell in skills) {
//                chosenSkillsS += spell.toString() + "#"
//            }
//            chosenSkillsS = chosenSkillsS.substring(0, chosenSkillsS.length - 1)
//        }
//
//        bw.write("skills=" + chosenSkillsS + "\n")
//        bw.write("provisionsValue=4\n")
//        bw.write("provisions=10\n")
//        bw.write("gold=0\n")
//    }
//
//    override fun validateCreationSpecificParameters(): String? {
//        val sb = StringBuilder()
//        if (this.skills.isEmpty()) {
//            sb.append(getString(R.string.chosenSkills))
//        }
//
//        return sb.toString()
//    }
//
//    fun addSkill(skill: MRSkill) {
//        if (!skills.contains(skill)) {
//            skills.add(skill)
//        }
//    }
//
//    fun removeSkill(skill: Int) {
//        skills.removeAt(skill)
//    }
//}
